Index: files/ultra96-power-button-check.c
===================================================================
--- files.orig/ultra96-power-button-check.c
+++ files/ultra96-power-button-check.c
@@ -44,7 +44,6 @@ static void power_button_pushed(void *);
 static void usage(char *);
 
 static char command[256]={0};
-const char *lcd_display= "python /sbin/groove-rgb-lcd.py";
 
 static void usage(char *cmd)
 {
@@ -57,7 +56,6 @@ static void power_button_pushed(void *du
 {
 
 	printf("Power button pressed. Powering down...\n");
-	(void) system(lcd_display);
 	(void) system(command);
 }
 
